test/functional/Valhalla/src/org/openj9/test/lworld/ValueTypeSystemArraycopyTests.java

```diff
@@ -0,0 +1,396 @@
+
+		Object[] dst = new Object[ARRAY_SIZE];
+
+		for (int i=0; i < ARRAY_SIZE; i++) {
+			dst[i] = new SomePrimitiveValueClass(i*11);
+		}
+
+		System.arraycopy(src, 0, dst, 0, ARRAY_SIZE);
+
+		checkResults(src, dst);
+	}
+
+	static private void testOBJIF(SomeInterface[] dst) {
+
+		Object[] src = new Object[ARRAY_SIZE];
+
+		for (int i=0; i < ARRAY_SIZE; i++) {
+			src[i] = new SomePrimitiveValueClass(i*11);
+		}
+
+		System.arraycopy(src, 0, dst, 0, ARRAY_SIZE);
+
+		checkResults(src, dst);
+	}
+
+	@Test(priority=1)
+	static public void testSystemArrayCopy1() throws Throwable {
+
+		initArrays();
+		testVTVT(vtArraySrc, vtArrayDst); // Fist invocation (Interpreter)
+
+		initArrays();
+		testVTVT(vtArraySrc, vtArrayDst);
+
+		checkResults(vtArraySrc, vtArrayDst);
+	}
+
+	@Test(priority=1)
+	static public void testSystemArrayCopy2() throws Throwable {
+
+		initArrays();
+		testVTIF(vtArraySrc, ifVtArrayDst); // Fist invocation (Interpreter)
+
+		initArrays();
+		testVTIF(vtArraySrc, ifVtArrayDst);
+
+		checkResults(vtArraySrc, ifVtArrayDst);
+	}
+
+	@Test(priority=1)
+	static public void testSystemArrayCopy3() throws Throwable {
+
+		initArrays();
+		testIFVT(ifVtArraySrc, vtArrayDst); // Fist invocation (Interpreter)
+
+		initArrays();
+		testIFVT(ifVtArraySrc, vtArrayDst);
+
+		checkResults(ifVtArraySrc, vtArrayDst);
+	}
+
+	@Test(priority=1)
+	static public void testSystemArrayCopy4() throws Throwable {
+
+		initArrays();
+		testIDID(idArraySrc, idArrayDst); // Fist invocation (Interpreter)
+
+		initArrays();
+		testIDID(idArraySrc, idArrayDst);
+
+		checkResults(idArraySrc, idArrayDst);
+	}
+
+	@Test(priority=1)
+	static public void testSystemArrayCopy5() throws Throwable {
+
+		initArrays();
+		testIDIF(idArraySrc, ifIdArrayDst); // Fist invocation (Interpreter)
+
+		initArrays();
+		testIDIF(idArraySrc, ifIdArrayDst);
+
+		checkResults(idArraySrc, ifIdArrayDst);
+	}
+
+	@Test(priority=1)
+	static public void testSystemArrayCopy6() throws Throwable {
+
+		initArrays();
+		testIFID(ifIdArraySrc, idArrayDst); // Fist invocation (Interpreter)
+
+		initArrays();
+		testIFID(ifIdArraySrc, idArrayDst);
+
+		checkResults(ifIdArraySrc, idArrayDst);
+	}
+
+	@Test(priority=1)
+	static public void testSystemArrayCopy7() throws Throwable {
+
+		initArrays();
+		testIFIF(ifIdArraySrc, ifIdArrayDst); // Fist invocation (Interpreter)
+
+		initArrays();
+		testIFIF(ifIdArraySrc, ifIdArrayDst);
+
+		checkResults(ifIdArraySrc, ifIdArrayDst);
+
+		initArrays();
+		testIFIF(ifVtArraySrc, ifVtArrayDst);
+
+		checkResults(ifVtArraySrc, ifVtArrayDst);
+
+		initArrays();
+		testIFIF(ifArray1, ifIdArrayDst);
+
+		checkResults(ifArray1, ifIdArrayDst);
+
+		initArrays();
+		testIFIF(ifIdArraySrc, ifArray1);
+
+		checkResults(ifIdArraySrc, ifArray1);
+
+		initArrays();
+		testIFIF(ifArray2, ifVtArrayDst);
+
+		checkResults(ifArray2, ifVtArrayDst);
+
+		initArrays();
+		testIFIF(ifVtArraySrc, ifArray2);
+
+		checkResults(ifVtArraySrc, ifArray2);
+	}
```

Comment:
It would be good to include cases where an `ArrayStoreException` or a `NullPointerException` should result.  For instance, if a source array whose component type is `SomeInterface` contains

-  a null reference or an instance of `SomeIdentityClass`, and the destination array has a component type of `SomePrimitiveValueClass`, or
- contains an instance of `SomePrimitiveValueClass` and the destination array has a component type of `SomeIdentityClass`

an exception should be thrown, but the destination array should still have been updated up to the point at which the "bad" element was encountered.

Reply:
Addressed in [02d9874](https://github.com/eclipse-openj9/openj9/compare/master...a7ehuo:openj9:add-test-systemarraycopy-PR).